### ISSUE TITLE
Script and Global Variables Are Visually Indistinguishable

### ISSUE THREAD

**Issue by peterstory** [jmoenig/Snap#2988]
Working on BJC Unit 2's Lab 1, one of my students created a very confusing bug:
https://snap.berkeley.edu/snap/snap.html#present:Username=peter_story&ProjectName=U2L1-NumberGuessingScriptBug

Reading through the code of the `numberguessinggame` block, everything seems fine! But the program has very strange behavior:
- The say block (added for debugging) doesn't say the random number
- The logic doesn't work properly. Try -1, 0, 20, and you'll see what I mean.

It turns out, the issue was because the student had created a `secret number` reporter block, and was using that, instead of the script variable. This seems like a UI design issue – when code is different, it should look different. Why not make script variables gray? 

The attached screenshot illustrates the problem.
<img width="1320" alt="Screen Shot 2022-02-25 at 3 20 01 PM" src="https://user-images.githubusercontent.com/1322135/155795252-7f8d2f3b-0996-4edc-ab6d-e3fd8b77a505.png">

**Comment by jguille2** [comment]
Hi Peter @peterstory ,

> Why not make script variables gray?

This is not a solution.
You can create gray (other category) reporters with the same name.

So, the "issue" is variables and reporters has the same shape.
But this is not so bad (I think). Because variables are also reporting... then it is a good visual metaphor.
And the problems with their (variables) names is also in their design. I mean that you can create a "script variable" with the same name of a "local/global variable". You can also create a reporter with the same name as a "primitive reporter". And then, their evaluation depends on the context!! And you have to take care of names!!

Joan

**Comment by brianharvey** [comment]
> when code is different, it should look different.

This isn't a principle in text languages!  You can have a dozen variables named x with different scopes if you want.  (Worse, in a sans-serif font you can't visually distinguish a variable named capital I from one named lower case l.)

As for variables and reporters, we _do_ have a visual difference: variables are orange.  If you make an orange custom reporter, presumably it's because you've _chosen_ to have it look like a variable, for whatever reason.

**Comment by DarDoro** [comment]
Function call is usually denoted by "()".
In Snap! the only way to distinguish "var x" from reporter "x" is looking at context menu ;)
Maybe something like ![untitled script pic - 2022-02-26T185123 805](https://user-images.githubusercontent.com/16109071/155854217-ac04fc81-61ca-4fd5-8d99-e3e4aeb7430b.png)

**Comment by peterstory** [comment]
I agree that it is possible to avoid this issue, but the problem when teaching is that students make all kinds of creative mistakes. Thus, the more usable the programming environment, the better. And this is clearly a usability issue. 

Typically, in a text-based language, when code references a variable named `secret_number`, the variable with the closest scope is used. Thus, in the example I gave, the script variable (i.e., instance variable) would be used. The usability issue is that this isn't what happens in Snap – if the reporter block is dragged in, that can be used. The reason this arose in this case was because the student was refactoring code that had originally been written outside the `numberguessinggame` block. This happened in a class of 20 students, so I can't imagine this occurrence would be very rare. 

So in this case, I think a text-based language would be *more* usable. The closest analogy with a text-based language would be if you had a hidden, non-printable character as part of a variable name. Then, you could get different behavior if you copy-pasted or retyped the variable name. Hence, why good text editors use fixed-width fonts and expand non-printable characters. 

I don't know what the best solution would be for this, but it doesn't seem like it should be an insurmountable problem.

**Comment by gigamonkey** [comment]
Wait! This means that Snap is a Lisp-2. The Scheme gods are likely to smite us all for this blasphemy. ;-) (I'm new here but I agree with @peterstory that needing to know where a thing was dragged from to know what it means is pretty confusing. Here's a highly simplified example:

![image](https://user-images.githubusercontent.com/250053/156279609-cb9ec52c-73b3-42fc-847b-232c2e041ed9.png)

In my actual script this block does not say 42 but rather this:

![image](https://user-images.githubusercontent.com/250053/156280004-8d3c7fb8-8064-4c15-8f62-af5359d368b0.png)

because the test in 

![image](https://user-images.githubusercontent.com/250053/156279753-2b97fbe3-9e44-4411-989f-0d1a7dc17522.png)

Is actually this block:

![image](https://user-images.githubusercontent.com/250053/156279902-fab8a3fa-e144-45e8-8984-07e9fe83a8b8.png)

I agree that not being able to visually distinguish those two cases is unfortunate. If Snap was a Lisp-1 and the test in the say was evaluated to be the script variable rather than the global block, then I think all would be well.
